### PR TITLE
Update Python versions used in CI workflow

### PR DESCRIPTION
**Pull request by c-bata**
## Motivation

Currently, lots of CI workflows uses Python 3.8, which is already EOL.

## Description of the changes

Update Python versions to 3.12 except for `sphinx-build.yml`.

### CODE REVIEW

**Comment by nabenabe0928** [comment]
@not522 Could you review this PR?

File: .github/workflows/mac-tests.yml
Comment (not522) [inline .github/workflows/mac-tests.yml:1]: Please update line 40 & 42.

File: .github/workflows/matplotlib-tests.yml
Comment (not522) [inline .github/workflows/matplotlib-tests.yml:1]: Please update line 35 & 37.

File: .github/workflows/speed-benchmarks.yml
Comment (not522) [inline .github/workflows/speed-benchmarks.yml:1]: Please update line 29 & 31.

File: .github/workflows/windows-tests.yml
Comment (not522) [inline .github/workflows/windows-tests.yml:1]: Please update line 40 & 42.

**Comment by c-bata** [comment]
@not522 Thank you for your review. I addressed your comments in cea73b9.

**Review by not522 — APPROVED** [review]
LGTM!

**Review by nabenabe0928 — APPROVED** [review]
LGTM!

**Pull request merged**